Add Granger

Alexander984 opened the pull request:
1. Add Python support to STEER
2. Add unit tests to support Python integration
3. Add Granger Causality
4. Update build files to support Granger

Alexander984 committed:
Added Granger

Alexander984 committed:
Added Python utilities and unit tests

Alexander984 committed:
Added Python utilities and unit tests

Alexander984 committed:
Added Python utilities and unit tests

Alexander984 committed:
More Python utilities and unit tests

Alexander984 committed:
Granger fully integrated first pass

Alexander984 committed:
Updated environment check to support Granger

Alexander984 committed:
Code quality

Alexander984 committed:
Updated validation results

Alexander984 committed:
Add Granger validation tag

Alexander984 committed:
Updated documentation

sillyvester commented on src/nist-sts/granger/granger.py:
```diff
@@ -0,0 +1,91 @@
+
+
+def granger_test(restricted_set: np.array, unrestricted_set: np.array,
+                 target: np.array) -> Tuple[float, float]:
```
Tuple[float, float] indicates that we only return 2 items in the tuple, but on line 73 we return 3 things. (I'm not sure if diff needs to be returned, i think lr and sf would be sufficient, unless i'm missing something

sillyvester commented on src/nist-sts/granger/granger.py:
```diff
@@ -0,0 +1,91 @@
+        sequences[0].append(bitstream[i:i + n])
+    sequences[0] = np.array(sequences[0])
+
+    restricted_set = np.array(sequences[1])
```
We've made one change here for experiments on our most recent paper. We add a constant term to both the restricted and unrestricted sets to ensure that the model isn't forced through the origin:

```
restricted_set = sm.add_constant(restricted_set)
unrestricted_set = `sm.add_constant(unrestricted_set)
```

artsamuilik replied:
Solid prediction from Josh. That's exactly what we've implemented

artsamuilik committed:
Update granger.py

Most importantly added constants to the restricted and unrestricted sets. Changed formatting for the import_bitstream function. Changed CLI argument for offset from number of offset bits to number of offset windows. Changed the help message and format_bitstream function accordingly.

artsamuilik committed:
Merge pull request #3 from SMU-DDI/add-granger-updates

Update granger.py

Most importantly added constants to the restricted and unrestricted sets. Changed formatting for the import_bitstream function. Changed CLI argument for offset from number of offset bits to number of offset windows. Changed the help message and format_bitstream function accordingly.

artsamuilik committed:
Update granger.py

Fixed the issue with receiving different p-values from granger.py outside of STEER versus in STEER. P-values now match Josh's granger code. (1 million bits are looked at)

artsamuilik committed:
Update STEER_Developer_Guide.md

Added a sentence that clarifies that ./build.sh must be run after editing the test files to make sure those changes are actually implemented in STEER.

artsamuilik committed:
granger.py - changed default window size to 8

artsamuilik committed:
Update expected_results_e.json

Updated expected validation results for e.bin in accordance with updated granger.py and a window size of 8 bits with offset of 6 windows

artsamuilik committed:
Update expected_results_pi.json

Updated expected validation results for pi.bin in accordance with updated granger.py and a window size of 8 bits with offset of 6 windows

artsamuilik committed:
Update expected_results_sha1.json

Updated expected validation results for sha1.bin in accordance with updated granger.py and a window size of 8 bits with offset of 6 windows

artsamuilik committed:
Update expected_results_sqrt2.json

Updated expected validation results for sqrt2.bin in accordance with updated granger.py and a window size of 8 bits with offset of 6 windows

artsamuilik committed:
Update expected_results_sqrt3.json

Updated expected validation results for sqrt3.bin in accordance with updated granger.py and a window size of 8 bits with offset of 6 windows

artsamuilik committed:
Granger Fully Integrated Parameter Updates

Updated parameters for e.bin so that granger validation runs faster and in accordance with the parameter change in the code (going from offset bits to offset windows). Specifically, changed the number of bits per window from 31 to 8, and corrected the number of offset bits from 186 to number of offset windows of 6.

artsamuilik committed:
Granger Fully Integrated Parameter Updates

Updated parameters for pi.bin so that granger validation runs faster and in accordance with the parameter change in the code (going from offset bits to offset windows). Specifically, changed the number of bits per window from 31 to 8, and corrected the number of offset bits from 186 to number of offset windows of 6.

artsamuilik committed:
Granger Fully Integrated Parameter Updates

Changed windows offset unit from "bits" to "windows"

artsamuilik committed:
Granger Fully Integrated Parameter Updates

Changed the correct section this time from units of "bits" to units of "windows"

artsamuilik committed:
Granger Fully Integrated Parameter Updates

Updated parameters for sha1.bin so that granger validation runs faster and in accordance with the parameter change in the code (going from offset bits to offset windows). Specifically, changed the number of bits per window from 31 to 8, and corrected the number of offset bits from 186 to number of offset windows of 6.

artsamuilik committed:
Granger Fully Integrated Parameter Updates

Updated parameters for sqrt2.bin so that granger validation runs faster and in accordance with the parameter change in the code (going from offset bits to offset windows). Specifically, changed the number of bits per window from 31 to 8, and corrected the number of offset bits from 186 to number of offset windows of 6.

artsamuilik committed:
Granger Fully Integrated Parameter Updates

Updated parameters for sqrt3.bin so that granger validation runs faster and in accordance with the parameter change in the code (going from offset bits to offset windows). Specifically, changed the number of bits per window from 31 to 8, and corrected the number of offset bits from 186 to number of offset windows of 6.

sillyvester committed:
minor notation changes
